internal/ls/signaturehelp.go

```diff
@@ -0,0 +1,1122 @@
+	if predicate != nil {
+		returnType.WriteString(c.TypePredicateToString(predicate))
+	} else {
+		returnType.WriteString(c.TypeToString(c.GetReturnTypeOfSignature(candidateSignature)))
+	}
+	return returnType.String()
+}
+
+func itemInfoForTypeParameters(candidateSignature *checker.Signature, c *checker.Checker, enclosingDeclaration *ast.Node, sourceFile *ast.SourceFile) []*signatureHelpItemInfo {
+	printer := printer.NewPrinter(printer.PrinterOptions{NewLine: core.NewLineKindLF}, printer.PrintHandlers{}, nil)
+
+	var typeParameters []*checker.Type
+	if candidateSignature.Target() != nil {
+		typeParameters = candidateSignature.Target().TypeParameters()
+	} else {
+		typeParameters = candidateSignature.TypeParameters()
+	}
+	var getTypeParameters []signatureHelpParameter = []signatureHelpParameter{}
+	for _, typeParameter := range typeParameters {
+		getTypeParameters = append(getTypeParameters, createSignatureHelpParameterForTypeParameter(typeParameter, sourceFile, enclosingDeclaration, c, printer))
+	}
+
+	thisParameter := []signatureHelpParameter{}
+	if candidateSignature.ThisParameter() != nil {
+		thisParameter = []signatureHelpParameter{createSignatureHelpParameterForParameter(candidateSignature.ThisParameter(), printer, sourceFile, c)}
+	}
+
+	// Creating display label
+	typeParameterDisplayParts := []string{}
+	for _, typeParameter := range getTypeParameters {
+		typeParameterDisplayParts = append(typeParameterDisplayParts, *typeParameter.parameterInfo.Label.String)
+	}
+
+	// Creating display parts for parameters. For example, <T>(a: string, b: number)
+	lists := c.GetExpandedParameters(candidateSignature, ptrTo(false))
+
+	var result []*signatureHelpItemInfo
+	parameterLabels := []string{}
+	for _, parameterList := range lists {
+		parameters := thisParameter
+		for _, param := range parameterList {
+			parameter := createSignatureHelpParameterForParameter(param, printer, sourceFile, c)
+			parameterLabels = append(parameterLabels, *parameter.parameterInfo.Label.String)
+			parameters = append(parameters, parameter)
+		}
+
+		result = append(result, &signatureHelpItemInfo{
+			isVariadic:   false,
+			parameters:   &getTypeParameters,
+			displayParts: []string{scanner.TokenToString(ast.KindLessThanToken), strings.Join(typeParameterDisplayParts, ", "), scanner.TokenToString(ast.KindGreaterThanToken), scanner.TokenToString(ast.KindOpenParenToken), strings.Join(parameterLabels, ", "), scanner.TokenToString(ast.KindCloseParenToken)},
+		})
+	}
+	return result
+}
+
+func itemInfoForParameters(candidateSignature *checker.Signature, c *checker.Checker, enclosingDeclaratipn *ast.Node, sourceFile *ast.SourceFile) []*signatureHelpItemInfo {
+	printer := printer.NewPrinter(printer.PrinterOptions{NewLine: core.NewLineKindLF}, printer.PrintHandlers{}, nil)
+
+	var getTypeParameters []signatureHelpParameter = []signatureHelpParameter{}
+	if candidateSignature.TypeParameters() != nil && len(candidateSignature.TypeParameters()) != 0 {
```

Comment:
This here is what I mentioned elsewhere that could become just `len(candidateSignature.TypeParameters()) != 0`.